R/plot-build.R

```diff
@@ -481,3 +398,132 @@ table_add_tag <- function(table, label, theme) {
     t = place$t, l = place$l, b = place$b, r = place$r
   )
 }
+
+# Add the legends to the gtable
+table_add_legends <- function(table, legends, theme) {
+
+  # Extract sizes
+  widths <- heights <- set_names(
+    rep(list(unit(0, "cm")), length(legends)),
+    names(legends)
+  )
+  if (is.zero(legends)) {
+    legends <- rep(list(zeroGrob()), 5)
+    names(legends) <- c(.trbl, "inside")
+  }
+  empty <- vapply(legends, is.zero, logical(1))
+  widths[!empty]  <- lapply(legends[!empty], gtable_width)
+  heights[!empty] <- lapply(legends[!empty], gtable_height)
+  spacing <- theme$legend.box.spacing %||% unit(0.2, "cm")
+
+  if (!is.zero(legends$inside)) {
+    # Set the justification of the legend box
+    just <- valid.just(calc_element("legend.justification.inside", theme))
+
+    position <- theme$legend.position.inside %||% just
+    legends$inside <- editGrob(
+      legends$inside,
+      vp = viewport(
+        x = position[1],
+        y = position[2],
+        just = just,
+        height = heights$inside,
+        width  = widths$inside
+      )
+    )
+  }
+
+  legends[.trbl] <- lapply(.trbl, function(position) {
+    box <- .subset2(legends, position)
+    if (is.zero(box)) {
+      return(box)
+    }
+    just <- paste0("legend.justification.", position)
+    just <- valid.just(calc_element(just, theme))
+
+    # First value is xjust, second value is yjust
+    xjust <- just[1]
+    yjust <- just[2]
+
+    margin <- margin(1 - yjust, 1 - xjust, yjust, xjust, unit = "null")
+
+    vp <- viewport(
+      x = xjust, y = yjust, just = just,
+      height = .subset2(heights, position),
+      width  = .subset2(widths,  position)
+    )
+    box <- editGrob(box, vp = vp)
+    gtable_add_padding(box, margin)
+  })
```

Comment:
For reasons I can't quite articulate, I feel like this code should be part of the guide-box building code instead of legend placement code.

Reply:
agreed - this feels like it belongs in `package_box()`